_posts/2017-11-30-totw-119.md

```diff
@@ -0,0 +1,264 @@
+`::example::util::foo`. This brittleness can be avoided by qualifying the used
+name fully: `using ::foo::Bar`.
+
+The only time that a relative name is unambiguous and cannot possibly be broken
+by outside declarations is if it refers to a name that is already inside your
+current namespace:
+
+```c++
+namespace example {
+namespace util {
+namespace internal {
+
+struct Params { /* ... */ };
+
+}  // namespace internal
+
+using internal::Params;  // OK, same as ::example::util::internal::Params
+```
+
+This follows the same logic that we discussed in the previous section.
+
+What if a name lives in a sibling namespace, such as `::example::tools::Thing`?
+You can say either `tools::Thing` or `::example::tools::Thing`. The fully
+qualified name is always correct, but it may also be appropriate to use the
+relative name. Use your own judgment.
+
+A cheap way to avoid a lot of these problems is not to use namespaces in your
+project that are the same as popular top-level namespaces (such as `util`); the
+Style Guide [recommends this practice]
+(http://engdoc/eng/doc/devguide/cpp/styleguide.shtml#Namespace_Names)
+explicitly.
+
+### Demo
+
+The following code shows examples of both failure modes.
+
+**helper.h:**
+
+```c++
+namespace bar {
+namespace foo {
+
+// ...
+
+}  // namespace foo
+}  // namespace bar
+```
+
+**some_feature.h:**
+
+```c++
+extern int f;
+```
+
+**Your code:**
+
+```c++
+#include "helper.h"
+#include "some_feature.h"
+
+namespace foo {
+void f();
+}  // namespace foo
+
+// Failure mode #1: Alias at a bad scope.
+// using foo::f;  // Error: redeclaration (because of "f" declared in some_feature.h)
+
+namespace bar {
+
+// Failure mode #2: Alias badly qualified.
+// using foo::f;  // Error: No "f" in namespace ::bar::foo (because that namespace was declared in helper.h)
+
+// The recommended way, robust in the face of unrelated declarations:
+using ::foo::f;  // OK
+
+void UseCase() { f(); }
+
+}  // namespace bar
+```
+
+## Unnamed namespaces
+
+A *using-declaration* placed in an unnamed namespace can be accessed from the
+enclosing namespace and vice versa. If you already have an unnamed namespace at
+the top of the file, prefer putting all aliases there. From within that unnamed
+namespace, you gain an extra little bit of robustness against clashing with
+something declared in the enclosing namespace.
+
+```c++
+namespace example {
+namespace util {
+
+namespace {
+
+// Put all using-declarations in here. Don't spread them over the file.
+using ::foo::Bar;
+using ::foo::Quz;
+
+// In here, Bar and Quz refer inalienably to your aliases.
+
+}  // namespace
+
+// Can use both Bar and Quz here too. (But don't declare any entities called Bar or Quz yourself now.)
+```
+
+## Non-aliased names
+
+So far we have been talking about local aliases for distant names. But what if
+we want to use names directly and not create aliases at all? Should we say
+`util::Status` or `::util::Status`?
```

Comment:
Change to some Abseil code?

Reply:
/shruggie - Nobody should be re-using / nesting absl:: so that's not a great example.  That's why I left it with util.

Reply:
ok